python/Ganga/Core/GangaRepository/SubJobXMLList.py

```diff
@@ -438,6 +441,9 @@ def flush(self):
                 subjob_data = self.__get_dataFile(str(index))
                 subjob_obj = self._cachedJobs[index]
 
+                if subjob_obj is subjob_obj._getRoot():
+                    raise GangaException(self, "Subjob parent not set correctly in flush.")
```

Comment:
This was the problem with the lock that your PR fixes. Basically, the flusher was trying to flush a subjob that was in _cachedObjs but didn't have it's parent set. This can happen (though very rarely) with the previous code as there was a small gap between adding the loaded subjob to the list and setting the parent. This exception should now NEVER happen but would at least be a very obvious red flag as to the problem if it was ever seen. If the flush continues with the object parent not set correctly you will probably get lockups which would be a lot harder to debug.